### PR TITLE
Better handling of `typing.Optional`

### LINKED ISSUE #40: Handling of `typing.Optional`
Consider a function

```python
def func(x: List[int], param: Optional[str] = None): ...
```
How should Sciline handle the `Optional[str]` parameter,

1. look for `str` and inject it if found, or
2. look for `Optional[str]` and inject it, or
3. both?

- Option 1.) would be useful when `str` is provided as a parameter or from a parameter table.
- Option 2.) would be useful if there is a provider
  ```python
  def maybe_str() -> Optional[str]: ...
  ```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #40.

Should we add top-level docs for this?

### CODE REVIEW

File: tests/pipeline_with_optional_test.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        pipeline[Union[int, float]] = 3  # type: ignore[index]
+    with pytest.raises(ValueError):
+        pipeline[Optional[int]] = 3  # type: ignore[index]
+
+
+def test_union_requirement_leads_to_UnsatisfiedRequirement() -> None:
+    def require_union(x: Union[int, float]) -> str:
+        return f'{x}'
+
+    pipeline = sl.Pipeline([require_union])
```
Comment (jl-wynen) [inline tests/pipeline_with_optional_test.py:38]: Can you add a provider for `int` and/or `float` so this actually tests something?

File: tests/pipeline_with_optional_test.py
Comment (jl-wynen) [inline tests/pipeline_with_optional_test.py:1]: Please add tests where there is an optional and a required dependency on some value. And check that it works for any order or providers and whether the value can be provided or not.

Comment (SimonHeybrock) [reply tests/pipeline_with_optional_test.py:1]: I added one test, not entirely sure what the other cases you are asking for are.

File: tests/pipeline_with_optional_test.py (hunk 1 above)
Comment (SimonHeybrock) [reply tests/pipeline_with_optional_test.py:38]: Good point, see update!

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**